### PR TITLE
add best practices/template analyzer tooling

### PR DESCRIPTION
**Pull request by gardnerjr**
## PR Checklist

Adds new scripts to the scripts folder that let you analyze template content for best practices warnings.

for example, if i do this, in this branch:

`scripts> npm run analyze-warnings-noloc ..\Workbooks`

it will analyze all of the workbook content in the Workbooks folder (recursively) and dump out all kinds of warnings (random sample, not to pick on anyone:

```
..\Workbooks\SapMonitor\SapHana\SapHana.workbook: Warning KQL003: Avoid Logs queries just doing print (use criteria/JSON params) [Step "param_memfulldate": // bring in the calculated values from parambag_tabledata let myparambag=parse_json('{parambag_tabl...]
..\Workbooks\SapMonitor\SapHana\SapHana.workbook: Warning KQL004: Avoid Logs queries just doing datatable (use JSON data source instead) [Step "param_check_table_exists": let checktable = (tb:string) { // 0 = no error // 1 = table does not exist // 2 = table exists, ...]
..\Workbooks\SapMonitor\SapHana\SapHana.workbook: Warning SIZE000: Overall JSON size large (Slow loading) [279182 ch > 100000]
..\Workbooks\SapMonitor\SapNetWeaver\SapNetWeaver.workbook: Warning KQL001: Should not have too many Logs queries (Avoid throttling) [127 queries > 40]
..\Workbooks\SapMonitor\SapNetWeaver\SapNetWeaver.workbook: Warning KQL002: Very large logs query (simplify?) [Step "param_check_table_exists": (3729 ch > 3000) let checktable = (tb:string) { // 0 = no error // 1 = table does not exist (need to enable RFC) ...]
..\Workbooks\Storage\Overview\Overview.workbook: Warning PARAM002: param "ResourceName" is a criteria parameter and has a value set in the template content [Step "parameters - 1": ResourceName]
..\Workbooks\Traffic Analytics\Unified\ApplicationPorts\ApplicationPorts.workbook: Warning KQL005: Avoid Logs queries just doing parameter only queries (hard for users to edit, won't work in pins) [Step "subscriptions": {SubscriptionQueryToRun}]
..\Workbooks\Traffic Analytics\Unified\ApplicationPorts\ApplicationPorts.workbook: Warning KQL005: Avoid Logs queries just doing parameter only queries (hard for users to edit, won't work in pins) [Step "resourceGroups": {RGQueryToRun}]
..\Workbooks\Traffic Analytics\Unified\NSG_ACLHits\NSG_ACLHits.workbook: Warning SIZE000: Overall JSON size large (Slow loading) [211269 ch > 100000]
..\Workbooks\Traffic Analytics\Unified\OverviewPage\OverviewPage.workbook: Warning KQL001: Should not have too many Logs queries (Avoid throttling) [71 queries > 40]
..\Workbooks\Traffic Analytics\Unified\OverviewPage\OverviewPage.workbook: Error SIZE001: Overall JSON size very large (Slow loading) [1024081 ch > 500000]
..\Workbooks\Virtual Machines - Performance Analysis\Performance Analysis for a Group of VMs (Perf)\Performance Analysis for a Group of VMs (Perf).workbook: Warning KQL003: Avoid Logs queries just doing print (use criteria/JSON params) [Step "aggregateLeftLabel": print "{Aggregate:label}"]
..\Workbooks\Virtual Machines - Performance Analysis\Performance Analysis for a Group of VMs (Perf)\Performance Analysis for a Group of VMs (Perf).workbook: Warning KQL003: Avoid Logs queries just doing print (use criteria/JSON params) [Step "aggregateRightValue": print "{Aggregate:value}"]
..\Workbooks\Virtual Machines - Performance Analysis\Performance Analysis for a Group of VMs (Perf)\Performance Analysis for a Group of VMs (Perf).workbook: Warning KQL003: Avoid Logs queries just doing print (use criteria/JSON params) [Step "aggregateRightLabel": print "{Aggregate:label}"]
```

or you can run

`scripts> npm run summarize ..\Workbooks`

and it will give you statistics of all kinds of stuff

```
STEPS001 - Count of each step type
----------------------------------
query           5051    (avg 7.87)
text            3445    (avg 5.37)
groups          2157    (avg 3.36)
parameters      1965    (avg 3.06)
metrics          960    (avg 1.50)
links/actions    640    (avg 1.00)


DATASOURCE001 - Count of each data source used in query steps
-------------------------------------------------------------
logs            3410    (avg 5.31)
arg              726    (avg 1.13)
adx              478    (avg 0.74)
merge            152    (avg 0.24)
json             140    (avg 0.22)
undefined         74    (avg 0.12)
arm               49    (avg 0.08)
azure health      11    (avg 0.02)
prom               3    (avg 0.00)
changes            3    (avg 0.00)
msgraph            2    (avg 0.00)
intune             2    (avg 0.00)
parameters         1    (avg 0.00)
```

### CODE REVIEW

File: scripts/src/Rules.ts [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+import { isEmpty, isNullOrWhitespace, notNullOrUndefined } from "./Utils";
+
+const JsonPath = require('jsonpath');
+
+// these regexs are copied here so this file can be reused in the github as well, self contained.
+/** regex for finding params in brackets, including any format specifiers, like :escape, which can be used in query steps */
+const ValidSpecifierRegex = "[_a-zA-Z0-9\xA0-\uFFFF\\-\\$\\@\\.\\[\\]\\*\\?\\(\\)\\<\\>\\=\\,\\:]*";
+const ValidParameterNameRegex = "[_a-zA-Z\xA0-\uFFFF][_a-zA-Z0-9\xA0-\uFFFF]*";
+//const ParameterRegex = new RegExp("\{" + ValidParameterNameRegex + "(:"+ ValidSpecifierRegex + ")?\}", "g");
+const OnlyParameterRegex = new RegExp("^\{" + ValidParameterNameRegex + "(:"+ ValidSpecifierRegex + ")?\}$", "g");
```
[the remark this answers is not in the document]
Comment (gardnerjr) [reply scripts/src/Rules.ts:13]: strange, the message *says* "escape sequence '{{`" but what it *means* is "escape sequence `\\{` ?

**Review by vutran01 — APPROVED** [review]
:shipit:

File: scripts/src/Rules.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gardnerjr) [reply scripts/src/Rules.ts]: fixed

[the remark this answers is not in the document]
Comment (gardnerjr) [reply scripts/src/Rules.ts]: fixed!

**Pull request merged**